gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DistributedRegion.java

```diff
@@ -1786,12 +1786,22 @@ private void cleanUpDestroyedTokensAndMarkGIIComplete(GIIStatus giiStatus)
 
         // in the first loop apply the committed/rolled back transactions
         if (!pendingTXStates.isEmpty()) {
+
+          final TXRegionState[] orderedTXRegionState = pendingTXStates
+              .toArray(new TXRegionState[pendingTXStates.size()]);
+
+          Arrays.sort(orderedTXRegionState, (t1, t2) -> {
+            final int o1 = System.identityHashCode(t1);
+            final int o2 = System.identityHashCode(t2);
+            return (o1 < o2) ? -1 : ((o1 == o2) ? 0 : 1);
+          });
+
```

Comment:
Better to use t1.getTXState().getTransactionId() and make TXId comparable that compares on memberId then uniqueId.